transactor/src/main/scala/io/mediachain/transactor/Copycat.scala

```diff
@@ -0,0 +1,186 @@
+          publishCommit(entry)
+          Xor.right(entry)
+        }
+      }
+    }
+
+    def update(commit: Commit[JournalUpdate]): Xor[JournalError, ChainEntry] = {
+      try {
+        updateChain(commit.operation)
+      } finally {
+        commit.release()
+      }
+    }
+    
+    private def updateChain(cmd: JournalUpdate): Xor[JournalError, ChainEntry] = {
+      val ref = cmd.ref
+      val cell = cmd.cell
+      index.get(ref) match {
+        case None => CommitError("invalid reference")
+        case Some(cref) => {
+          cref match {
+            case EntityReference(chain) => {
+              cell match {
+                case EntityChainCell(entity, xchain, meta) => {
+                  if (checkUpdate(ref, chain, entity, xchain)) {
+                    val newcell = EntityChainCell(ref, chain, meta)
+                    val newchain = datastore.put(newcell)
+                    index.put(ref, EntityReference(Some(newchain)))
+                    
+                    val entry = ChainEntry(nextSeqno(), ref, newchain, chain)
+                    publishCommit(entry)
+                    Xor.right(entry)
+                  } else CommitError("invalid chain cell")
+                }
+                case _ => CommitError("invalid chain")
+              }
+            }
+            case ArtefactReference(chain) => {
+              cell match {
+                case ArtefactChainCell(artefact, xchain, meta) => {
+                  if (checkUpdate(ref, chain, artefact, xchain)) {
+                    val newcell = ArtefactChainCell(ref, chain, meta)
+                    val newchain = datastore.put(newcell)
+                    index.put(ref, ArtefactReference(Some(newchain)))
+                    
+                    val entry = ChainEntry(nextSeqno(), ref, newchain, chain)
+                    publishCommit(entry)
+                    Xor.right(entry)
+                  } else CommitError("invalid chain cell")
+                }
+                case _ => CommitError("invalid chain")
+              }
+            }
+          }
+        }
+      }
+    }
+
+    def lookup(commit: Commit[JournalLookup]): Option[Reference] = {
+      try {
+        index.get(commit.operation.ref) match {
+          case Some(cref) => cref.chain
+          case None => None
+        }
+      } finally {
+        commit.release()
+      }
+    }
+
+    // helpers
+    private def nextSeqno() = {
+      val next = seqno
+      seqno += 1
+      next
+    }
+    
+
+    private def publishCommit(entry: JournalEntry) {
+      val event = JournalCommitEvent(entry)
+      clients.foreach(_.publish("journal-commit", event))
+    }
+    
+    
+    private def checkUpdate(ref: Reference, chain: Option[Reference],
+                            xref: Reference, xchain: Option[Reference]) = {
+      (xref == ref) && ((xchain == None) || (xchain == chain))
+    }
+        
+    // Snapshottable
+    override def install(reader: SnapshotReader) {
```

Comment:
On second thought, maybe leave the `override` in there so the compiler complains if the parent interface changes